GVFS/GVFS.Common/NuGetUpgrade/NuGetUpgrader.cs

```diff
@@ -0,0 +1,578 @@
+        }
+
+        private void UnzipPackage()
+        {
+            ZipFile.ExtractToDirectory(this.DownloadedPackagePath, this.ExtractedInstallerPath);
+        }
+
+        private bool TryRecursivelyDeleteInstallerDirectory(out string error)
+        {
+            error = null;
+            Exception e;
+            if (!this.fileSystem.TryDeleteDirectory(this.ExtractedInstallerPath, out e))
+            {
+                if (e != null)
+                {
+                    this.TraceException(
+                        e,
+                        nameof(this.TryRecursivelyDeleteInstallerDirectory),
+                        $"Exception encountered while deleting {this.ExtractedInstallerPath}.");
+                }
+
+                error = e?.Message ?? "Failed to delete directory, but no error was specified.";
+                return false;
+            }
+
+            return true;
+        }
+
+        public class NuGetUpgraderConfig
+        {
+            private readonly ITracer tracer;
+            private readonly LocalGVFSConfig localConfig;
+
+            public NuGetUpgraderConfig(ITracer tracer, LocalGVFSConfig localGVFSConfig)
+            {
+                this.tracer = tracer;
+                this.localConfig = localGVFSConfig;
+            }
+
+            public NuGetUpgraderConfig(
+                ITracer tracer,
+                LocalGVFSConfig localGVFSConfig,
+                string feedUrl,
+                string packageFeedName)
+                : this(tracer, localGVFSConfig)
+            {
+                this.FeedUrl = feedUrl;
+                this.PackageFeedName = packageFeedName;
+            }
+
+            public string FeedUrl { get; private set; }
+            public string PackageFeedName { get; private set; }
+            public string CertificateFingerprint { get; private set; }
+
+            /// <summary>
+            /// Check if the NuGetUpgrader is ready for use. A
+            /// NuGetUpgrader is considered ready if all required
+            /// config settings are present.
+            /// </summary>
+            public bool IsReady(out string error)
+            {
+                if (string.IsNullOrEmpty(this.FeedUrl) ||
+                    string.IsNullOrEmpty(this.PackageFeedName))
+                {
+                    error = string.Join(
+                        Environment.NewLine,
+                        "One or more required settings for NuGetUpgrader are missing.",
+                        $"Use `gvfs config [{GVFSConstants.LocalGVFSConfig.UpgradeFeedUrl} | {GVFSConstants.LocalGVFSConfig.UpgradeFeedPackageName}] <value>` to set the config.");
+                    return false;
+                }
+
+                error = null;
+                return true;
+            }
+
+            /// <summary>
+            /// Check if the NuGetUpgrader is configured.
+            /// </summary>
+            public bool IsConfigured(out string error)
```

Comment:
I'm not sure I understand the distinction between IsReady and IsConfigured. They both check for the existence of two settings, and both give very similar errors if both are not present.

Can we just go with `IsConfigured`? I prefer that name. Also, it would be nice to be more explicit about which expected setting(s) was/were not found.